pilot/pkg/config/mcp/controller.go

```diff
@@ -0,0 +1,248 @@
+	if !exists {
+		return nil, false
+	}
+	config := out.(model.Config)
+
+	return &config, true
+}
+
+func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+	_, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return nil, errors.New(fmt.Sprintf("List: unknown type %s", typ))
+	}
+	data, exists := c.configStore[typ]
+	if !exists {
+		c.logger.Infof("List: config not found for the type %s", typ)
+		return nil, nil
+	}
+	out := make([]model.Config, 0, len(c.configStore[typ]))
+	if namespace == "" {
+		for _, ns := range data {
+			ns.Range(func(key, value interface{}) bool {
+				out = append(out, value.(model.Config))
+				return true
+			})
+		}
+	} else {
+		ns, exists := data[namespace]
+		if !exists {
+			return nil, nil
+		}
+		ns.Range(func(key, value interface{}) bool {
+			out = append(out, value.(model.Config))
+			return true
+		})
+	}
+	return out, nil
+}
+
+func (c *Controller) Update(config model.Config) (newRevision string, err error) {
+	typ := config.Type
+	schema, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return "", errors.New(fmt.Sprintf("Update: unknown type %s", typ))
+	}
+	if err := schema.Validate(config.Name, config.Namespace, config.Spec); err != nil {
+		return "", err
+	}
+
+	ns, exists := c.configStore[typ][config.Namespace]
+	if !exists {
+		return "", errNotFound
+	}
+
+	oldConfig, exists := ns.Load(config.Name)
+	if !exists {
+		return "", errNotFound
+	}
+
+	if config.ResourceVersion != oldConfig.(model.Config).ResourceVersion {
+		return "", errors.New("old revision")
+	}
+
+	rev := time.Now().String()
+	config.ResourceVersion = rev
+	ns.Store(config.Name, config)
+	return rev, nil
+}
+
+func (c *Controller) Create(config model.Config) (revision string, err error) {
+	typ := config.Type
+	schema, ok := c.ConfigDescriptor().GetByType(typ)
+	if !ok {
+		return "", errors.New("unknown type")
+	}
+	if err := schema.Validate(config.Name, config.Namespace, config.Spec); err != nil {
+		return "", err
+	}
+	ns, exists := c.configStore[typ][config.Namespace]
+	if !exists {
+		ns = new(sync.Map)
+		c.configStore[typ][config.Namespace] = ns
+	}
+
+	_, exists = ns.Load(config.Name)
+
+	if !exists {
+		tnow := time.Now()
+		config.ResourceVersion = tnow.String()
+
+		// Set the creation timestamp, if not provided.
+		if config.CreationTimestamp.Time.IsZero() {
+			config.CreationTimestamp.Time = tnow
+		}
+
+		ns.Store(config.Name, config)
+		return config.ResourceVersion, nil
+	}
+	return "", errAlreadyExists
+}
+
+func (c *Controller) Apply(change *mcpclient.Change) error {
+	for _, obj := range change.Objects {
+		descriptor, ok := c.descriptorsByMessageName[change.MessageName]
+		if !ok {
+			return fmt.Errorf("Apply: type not supported %s", change.MessageName)
+		}
+
+		c.updateCh <- changeEvent{
+			metaName:       obj.Metadata.Name,
+			resource:       obj.Resource,
+			descriptorType: descriptor.Type,
+		}
+	}
+	return nil
+}
+
+func (c *Controller) HasSynced() bool {
```

Comment:
Unlike the other methods however - this is very important to implement, should return true if MCP returned one of each watched objects ( and we need to make sure an empty result is sent ). This is used for readiness.

Reply:
@costinm That's a great feedback, I wasn't too sure what should be the logic for `HasSynced`, Also can you elaborate a bit more on your comment above "and we need to make sure an empty result is sent".

Reply:
Pilot will send requests for each resource type, we need to make sure MCP responds with an empty result for each resource type, so we can set 'hasSynced'. Probably it does - but need to double check, otherwise we'll keep waiting.

Reply:
The MCP server layer doesn't push an initial empty MCP response today. https://github.com/istio/istio/issues/7947 tracks this feature request.